src/MemoryTracing/include/MemoryTracing/MemoryInfoProducer.h

```diff
@@ -9,44 +9,52 @@
 
 #include <thread>
 
+#include "GrpcProtos/Constants.h"
 #include "MemoryTracing/MemoryInfoListener.h"
-#include "capture.pb.h"
 
 namespace orbit_memory_tracing {
+using MemoryInfoProducerRunFn = std::function<void(MemoryInfoListener*, int32_t)>;
 
-// This class periodically produces the SystemMemoryUsage information retrieved from /proc/meminfo.
-// When the MemoryInfoProducer is started, it creates a thread to extract the memory usage
-// information with a sampling interval specified in sampling_period_ns_, and send the extracted
-// information to the listener.
+// This class periodically produces the memory usage information. When the MemoryInfoProducer is
+// started, it creates a thread to extract the memory usage information with a sampling interval
+// specified in sampling_period_ns_, and send the extracted information to the listener.
 class MemoryInfoProducer {
  public:
-  explicit MemoryInfoProducer(uint64_t memory_sampling_period_ns)
-      : sampling_period_ns_(memory_sampling_period_ns) {}
+  explicit MemoryInfoProducer(uint64_t sampling_period_ns, int32_t pid, MemoryInfoProducerRunFn fn)
+      : sampling_period_ns_(sampling_period_ns), pid_(pid), producer_run_fn_(std::move(fn)) {}
 
   MemoryInfoProducer(const MemoryInfoProducer&) = delete;
   MemoryInfoProducer& operator=(const MemoryInfoProducer&) = delete;
   MemoryInfoProducer(MemoryInfoProducer&&) = delete;
   MemoryInfoProducer& operator=(MemoryInfoProducer&&) = delete;
 
-  ~MemoryInfoProducer() { Stop(); }
-
   void SetListener(MemoryInfoListener* listener) { listener_ = listener; }
+  void SetThreadName(std::string_view thread_name) { thread_name_ = thread_name; }
 
   void Start();
   void Stop();
 
  private:
   void SetExitRequested(bool exit_requested);
   void Run();
-  void ProduceSystemMemoryUsageAndSendToListener();
 
   uint64_t sampling_period_ns_;
   MemoryInfoListener* listener_ = nullptr;
+  int32_t pid_ = orbit_grpc_protos::kMissingInfo;
   bool exit_requested_ = true;
   absl::Mutex exit_requested_mutex_;
   std::unique_ptr<std::thread> thread_;
+  std::string thread_name_ = "MemInfoPr::Run";
+  MemoryInfoProducerRunFn producer_run_fn_;
 };
 
+// This function collect `SystemMemoryUsage` and send to listener.
+void SystemMemoryInfoProducerRunFn(MemoryInfoListener* listener, int32_t /*pid*/);
+// This function collect `CGroupMemoryUsage` and send to listener.
+void CGroupMemoryInfoProducerRunFn(MemoryInfoListener* listener, int32_t pid);
+// This function collect `ProcessMemoryUsage` and send to listener.
+void ProcessMemoryInfoProducerRunFn(MemoryInfoListener* listener, int32_t pid);
+
```

Comment:
If you define factory methods you can hide these from the user.

Reply:
Done. Thanks for the suggestion!